storage/badger/operation/reader_batch_writer.go

```diff
@@ -0,0 +1,130 @@
+package operation
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+)
+
+type ReaderBatchWriter struct {
+	db    *badger.DB
+	batch *badger.WriteBatch
+
+	addingCallback sync.Mutex // protect callbacks
+	callbacks      []func(error)
+}
+
+var _ storage.BadgerReaderBatchWriter = (*ReaderBatchWriter)(nil)
+
+func (b *ReaderBatchWriter) ReaderWriter() (storage.Reader, storage.Writer) {
+	// reusing the same underlying object, but expose with different interfaces
+	return b, b
+}
+
+func (b *ReaderBatchWriter) BadgerWriteBatch() *badger.WriteBatch {
+	return b.batch
+}
+
+func (b *ReaderBatchWriter) AddCallback(callback func(error)) {
+	b.addingCallback.Lock()
+	defer b.addingCallback.Unlock()
+
+	b.callbacks = append(b.callbacks, callback)
+}
+
+func (b *ReaderBatchWriter) Commit() error {
+	err := b.batch.Flush()
+
+	b.notifyCallbacks(err)
+
+	return err
+}
+
+func (b *ReaderBatchWriter) notifyCallbacks(err error) {
+	b.addingCallback.Lock()
+	defer b.addingCallback.Unlock()
+
+	for _, callback := range b.callbacks {
+		callback(err)
+	}
+}
+
+func WithReaderBatchWriter(db *badger.DB, fn func(storage.BadgerReaderBatchWriter) error) error {
+	batch := NewReaderBatchWriter(db)
+
+	err := fn(batch)
+	if err != nil {
+		// fn might use lock to ensure concurrent safety while reading and writing data
+		// and the lock is usually released by a callback.
+		// in other words, fn might hold a lock to be released by a callback,
+		// we need to notify the callback for the locks to be released before
+		// returning the error.
+		batch.notifyCallbacks(err)
+		return err
+	}
+
+	return batch.Commit()
+}
+
+func NewReaderBatchWriter(db *badger.DB) *ReaderBatchWriter {
+	return &ReaderBatchWriter{
+		db:    db,
+		batch: db.NewWriteBatch(),
+	}
+}
+
+// ToReader is a helper function to convert a BadgerReaderBatchWriter to a Reader
```

Comment:
```suggestion
// ToReader is a helper function to convert a *badger.DB to a Reader
```
I think? The input is a DB, so this makes more sense to me.